<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fetch statistics of duplicated fragments — get_dupl_stats_from_bams • nomeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fetch statistics of duplicated fragments — get_dupl_stats_from_bams" />
<meta property="og:description" content="Fetch statistics of duplicated fragments" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nomeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fetch statistics of duplicated fragments</h1>
    
    <div class="hidden name"><code>get_dupl_stats_from_bams.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fetch statistics of duplicated fragments</p>
    </div>

    <pre class="usage"><span class='fu'>get_dupl_stats_from_bams</span><span class='op'>(</span>
  <span class='va'>bamfiles</span>,
  <span class='va'>samplenames</span>,
  <span class='va'>regions</span>,
  <span class='va'>genome</span>,
  collapseBySample <span class='op'>=</span> <span class='cn'>TRUE</span>,
  mapqMin <span class='op'>=</span> <span class='fl'>0</span>,
  mapqMax <span class='op'>=</span> <span class='fl'>255</span>,
  max_read_size <span class='op'>=</span> <span class='fl'>1000L</span>,
  ncores <span class='op'>=</span> <span class='fl'>1L</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bamfiles</th>
      <td><p><code>character</code> paths to bam files</p></td>
    </tr>
    <tr>
      <th>samplenames</th>
      <td><p><code>character</code> names of samples. If <code>collapseBySample</code> is <code>TRUE</code> counts are aggregated across bam files with same sample name.</p></td>
    </tr>
    <tr>
      <th>regions</th>
      <td><p>GRanges object for regions of interest.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p><code>character</code> path to fasta file which was used as reference for alignment.</p></td>
    </tr>
    <tr>
      <th>collapseBySample</th>
      <td><p><code>logical</code> indicating whether to collapse counts for bam files with same <code>samplenames</code>. If <code>FALSE</code> prefix <code>s</code> followed by index is added to <code>samplenames</code>.</p></td>
    </tr>
    <tr>
      <th>mapqMin</th>
      <td><p><code>integer</code> setting minimum MAPQ of reads to be included into analysis.</p></td>
    </tr>
    <tr>
      <th>mapqMax</th>
      <td><p><code>integer</code> setting maximum MAPQ of reads to be included into analysis.</p></td>
    </tr>
    <tr>
      <th>max_read_size</th>
      <td><p>maximum read length.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>number of cores to use.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>tibble</code> where each row corresponds to sample - region combination. 
Columns of the returned <code>tibble</code> represent:</p><dl>
  <dt>SampleName</dt><dd><p>Name of a sample as defined by <code>samplenames</code>.</p></dd>
  <dt>seqnames, start, end</dt><dd><p>seqnames (or chromosomes) and coordinates of corresponding regions.</p></dd>
  <dt>names</dt><dd><p>the same as <code><a href='https://rdrr.io/r/base/names.html'>names(regions)</a></code> if it is not <code>NULL</code> or index of a corresponding region in <code>regions</code>.</p></dd>
  <dt>nFragsFetched</dt><dd><p>number of fetched fragments before filtering.</p></dd>
  <dt>nFragsNonUnique</dt><dd><p>number of non-unique fragments as defined by coordinates and methylation profiles.</p></dd>
  <dt>duplStatsMatrix</dt><dd><p><code>matrix</code> with 2 columns. First column represents (<code>n_uniq_aln</code>) number of unique alignments after deduplication and second 
  column (<code>n_times_duplicated</code>) represents number of times each of the alignments was duplicated.
  For example, row with numbers 100 at the first column and 1 at the second column should be interpreted as following: 100 alignments in the data has exactly 1 unique fragment.
  In other words, 100 fragments are unique.
  Another example, row with numbers 200 at the first column and 5 at the second column means that before deduplication 5 alignments were supported by 200 duplicated fragments each.</p></dd>
  <dt>duplFragNamesList</dt><dd><p><code>list</code> of vectors containing names of duplicated fragments. Names of list elements are encodings of unique alignments and can be ignored.</p></dd>

</dl>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Evgeniy A. Ozonov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


